File: rust/src/values.rs
use std::collections::HashMap;
use std::error;
use std::fmt;
use std::io;
use std::hash::Hash;


pub const TYPE_NIL: u8 = 0;
pub const TYPE_U8: u8 = 1;
pub const TYPE_I8: u8 = 2;
pub const TYPE_U16: u8 = 3;
pub const TYPE_I16: u8 = 4;
pub const TYPE_U32: u8 = 5;
pub const TYPE_I32: u8 = 6;
pub const TYPE_U64: u8 = 7;
pub const TYPE_I64: u8 = 8;
pub const TYPE_FLOAT: u8 = 9;
pub const TYPE_STR: u8 = 10;
pub const TYPE_STR_IDX: u8 = 11;
pub const TYPE_RAW: u8 = 12;
pub const TYPE_ARR: u8 = 13;
pub const TYPE_MAP: u8 = 14;

pub const STR_TYPE_NIL: &'static str = "nil";
pub const STR_TYPE_U8: &'static str = "u8";
pub const STR_TYPE_I8: &'static str = "i8";
pub const STR_TYPE_U16: &'static str = "u16";
pub const STR_TYPE_I16: &'static str = "i16";
pub const STR_TYPE_U32: &'static str = "u32";
pub const STR_TYPE_I32: &'static str = "i32";
pub const STR_TYPE_U64: &'static str = "u64";
pub const STR_TYPE_I64: &'static str = "i64";
pub const STR_TYPE_FLOAT: &'static str = "float";
pub const STR_TYPE_STR: &'static str = "str";
pub const STR_TYPE_STR_IDX: &'static str = "str_idx";
pub const STR_TYPE_RAW: &'static str = "raw";
pub const STR_TYPE_ARR: &'static str = "arr";
pub const STR_TYPE_MAP: &'static str = "map";

#[derive(PartialEq, Clone)]
pub enum Value {
    Nil,
    U8(u8),
    I8(i8),
    U16(u16),
    I16(i16),
    U32(u32),
    I32(i32),
    U64(u64),
    I64(i64),
    Float(f32),
    Str(String),
    Raw(Vec<u8>),
    Arr(Vec<Value>),
    Map(HashMap<Value, Value>),
}

impl Eq for Value {
}


impl Hash for Value {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        core::mem::discriminant(self).hash(state);
    }
}

impl fmt::Debug for Value {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            Value::Nil => write!(fmt, "nil"),
            Value::U8(val) => write!(fmt, "u8({:?})", val),
            Value::I8(val) => write!(fmt, "i8({:?})", val),
            Value::U16(val) => write!(fmt, "u16({:?})", val),
            Value::I16(val) => write!(fmt, "i16({:?})", val),
            Value::U32(val) => write!(fmt, "u32({:?})", val),
            Value::I32(val) => write!(fmt, "i32({:?})", val),
            Value::U64(val) => write!(fmt, "u64({:?})", val),
            Value::I64(val) => write!(fmt, "i64({:?})", val),
            Value::Float(val) => write!(fmt, "float({:?})", val),
            Value::Str(ref val) => write!(fmt, "str({:?})", val),
            Value::Raw(ref val) => write!(fmt, "str({:?})", val),
            Value::Arr(ref val) => write!(fmt, "arr({:?})", val),
            Value::Map(ref val) => write!(fmt, "str({:?})", val),
        }
    }
}

impl From<u8> for Value {
    fn from(val: u8) -> Value {
        Value::U8(val)
    }
}

impl From<i8> for Value {
    fn from(val: i8) -> Value {
        Value::I8(val)
    }
}

impl From<u16> for Value {
    fn from(val: u16) -> Value {
        Value::U16(val)
    }
}

impl From<i16> for Value {
    fn from(val: i16) -> Value {
        Value::I16(val)
    }
}

impl From<u32> for Value {
    fn from(val: u32) -> Value {
        Value::U32(val)
    }
}

impl From<i32> for Value {
    fn from(val: i32) -> Value {
        Value::I32(val)
    }
}

impl From<f32> for Value {
    fn from(val: f32) -> Value {
        Value::Float(val)
    }
}

impl From<String> for Value {
    fn from(val: String) -> Value {
        Value::Str(val)
    }
}

impl From<Vec<u8>> for Value {
    fn from(val: Vec<u8>) -> Value {
        Value::Raw(val)
    }
}

impl From<Vec<Value>> for Value {
    fn from(val: Vec<Value>) -> Value {
        Value::Arr(val)
    }
}

impl From<HashMap<Value, Value>> for Value {
    fn from(val: HashMap<Value, Value>) -> Value {
        Value::Map(val)
    }
}

impl Into<u8> for Value {
    fn into(self) -> u8 {
        match self {
            Value::U8(val) => val,
            _ => panic!("into error type {}", get_name_by_type(get_type_by_value(&self))),
        }
    }
}

impl Into<i8> for Value {
    fn into(self) -> i8 {
        match self {
            Value::I8(val) => val,
            _ => panic!("into error type {}", get_name_by_type(get_type_by_value(&self))),
        }
    }
}

impl Into<u16> for Value {
    fn into(self) -> u16 {
        match self {
            Value::U16(val) => val,
            _ => panic!("into error type {}", get_name_by_type(get_type_by_value(&self))),
        }
    }
}

impl Into<i16> for Value {
    fn into(self) -> i16 {
        match self {
            Value::I16(val) => val,
            _ => panic!("into error type {}", get_name_by_type(get_type_by_value(&self))),
        }
    }
}

impl Into<u32> for Value {
    fn into(self) -> u32 {
        match self {
            Value::U32(val) => val,
            _ => panic!("into error type {}", get_name_by_type(get_type_by_value(&self))),
        }
    }
}

impl Into<i32> for Value {
    fn into(self) -> i32 {
        match self {
            Value::I32(val) => val,
            _ => panic!("into error type {}", get_name_by_type(get_type_by_value(&self))),
        }
    }
}

impl Into<f32> for Value {
    fn into(self) -> f32 {
        match self {
            Value::Float(val) => val,
            _ => panic!("into error type {}", get_name_by_type(get_type_by_value(&self))),
        }
    }
}

impl Into<String> for Value {
    fn into(self) -> String {
        match self {
            Value::Str(val) => val,
            _ => panic!("into error type {}", get_name_by_type(get_type_by_value(&self))),
        }
    }
}

impl Into<Vec<u8>> for Value {
    fn into(self) -> Vec<u8> {
        match self {
            Value::Raw(val) => val,
            _ => panic!("into error type {}", get_name_by_type(get_type_by_value(&self))),
        }
    }
}

impl Into<Vec<Value>> for Value {
    fn into(self) -> Vec<Value> {
        match self {
            Value::Arr(val) => val,
            _ => panic!("into error type {}", get_name_by_type(get_type_by_value(&self))),
        }
    }
}

impl Into<HashMap<Value, Value>> for Value {
    fn into(self) -> HashMap<Value, Value> {
        match self {
            Value::Map(val) => val,
            _ => panic!("into error type {}", get_name_by_type(get_type_by_value(&self))),
        }
    }
}


pub fn get_type_by_value(value: &Value) -> u8 {
    match *value {
        Value::U8(_) => TYPE_U8,
        Value::I8(_) => TYPE_I8,
        Value::U16(_) => TYPE_U16,
        Value::I16(_) => TYPE_I16,
        Value::U32(_) => TYPE_U32,
        Value::I32(_) => TYPE_I32,
        Value::Float(_) => TYPE_FLOAT,
        Value::Str(_) => TYPE_STR,
        Value::Raw(_) => TYPE_RAW,
        Value::Arr(_) => TYPE_ARR,
        Value::Map(_) => TYPE_MAP,
        _ => TYPE_NIL,
    }
}

pub fn get_type_by_name(name: &str) -> u8 {
    match name {
        STR_TYPE_NIL => TYPE_NIL,
        STR_TYPE_U8 => TYPE_U8,
        STR_TYPE_I8 => TYPE_I8,
        STR_TYPE_U16 => TYPE_U16,
        STR_TYPE_I16 => TYPE_I16,
        STR_TYPE_U32 => TYPE_U32,
        STR_TYPE_I32 => TYPE_I32,
        STR_TYPE_FLOAT => TYPE_FLOAT,
        STR_TYPE_STR => TYPE_STR,
        STR_TYPE_RAW => TYPE_RAW,
        STR_TYPE_ARR => TYPE_ARR,
        STR_TYPE_MAP => TYPE_MAP,
        _ => TYPE_NIL,
    }
}

pub fn get_name_by_type(index: u8) -> &'static str {
    match index {
        TYPE_NIL => STR_TYPE_NIL,
        TYPE_U8 => STR_TYPE_U8,
        TYPE_I8 => STR_TYPE_I8,
        TYPE_U16 => STR_TYPE_U16,
        TYPE_I16 => STR_TYPE_I16,
        TYPE_U32 => STR_TYPE_U32,
        TYPE_I32 => STR_TYPE_I32,
        TYPE_FLOAT => STR_TYPE_FLOAT,
        TYPE_STR => STR_TYPE_STR,
        TYPE_RAW => STR_TYPE_RAW,
        TYPE_ARR => STR_TYPE_ARR,
        TYPE_MAP => STR_TYPE_MAP,
        _ => STR_TYPE_NIL,
    }
}
/// An enum of all error kinds.
#[derive(PartialEq, Eq, Copy, Clone, Debug)]
pub enum ErrorKind {
    /// will read bytes over than left bytes
    NoLeftSpaceError,
    /// the buffer over max
    BufferOverMaxError,
    /// The type not match
    TypeNotMatchError,
    /// the buffer can't parse the right data
    ParseError,
    /// miss the major data
    MissingError,
    /// string format must be utf-8
    StringFormatError,
    /// This kind is returned if the redis error is one that is
    /// not native to the system.  This is usually the case if
    /// the cause is another error.
    IoError,
    /// An extension error.  This is an error created by the server
    /// that is not directly understood by the library.
    ExtensionError,
}

#[derive(Debug)]
enum ErrorRepr {
    WithDescription(ErrorKind, &'static str),
    WithDescriptionAndDetail(ErrorKind, &'static str, String),
    ExtensionError(String, String),
    IoError(io::Error),
}

/// Represents a redis error.  For the most part you should be using
/// the Error trait to interact with this rather than the actual
/// struct.
pub struct RpError {
    repr: ErrorRepr,
}

/// Library generic result type.
pub type RpResult<T> = Result<T, RpError>;


impl PartialEq for RpError {
    fn eq(&self, other: &RpError) -> bool {
        match (&self.repr, &other.repr) {
            (&ErrorRepr::WithDescription(kind_a, _), &ErrorRepr::WithDescription(kind_b, _)) => {
                kind_a == kind_b
            }
            (&ErrorRepr::WithDescriptionAndDetail(kind_a, _, _),
             &ErrorRepr::WithDescriptionAndDetail(kind_b, _, _)) => kind_a == kind_b,
            (&ErrorRepr::ExtensionError(ref a, _), &ErrorRepr::ExtensionError(ref b, _)) => {
                *a == *b
            }
            _ => false,
        }
    }
}

impl From<io::Error> for RpError {
    fn from(err: io::Error) -> RpError {
        RpError { repr: ErrorRepr::IoError(err) }
    }
}


impl From<(ErrorKind, &'static str)> for RpError {
    fn from((kind, desc): (ErrorKind, &'static str)) -> RpError {
        RpError { repr: ErrorRepr::WithDescription(kind, desc) }
    }
}

impl From<(ErrorKind, &'static str, String)> for RpError {
    fn from((kind, desc, detail): (ErrorKind, &'static str, String)) -> RpError {
        RpError { repr: ErrorRepr::WithDescriptionAndDetail(kind, desc, detail) }
    }
}

impl error::Error for RpError {
    fn cause(&self) -> Option<&dyn error::Error> {
        match self.repr {
            ErrorRepr::IoError(ref err) => Some(err as &dyn error::Error),
            _ => None,
        }
    }
}

impl fmt::Display for RpError {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        match self.repr {
            ErrorRepr::WithDescription(_, desc) => desc.fmt(f),
            ErrorRepr::WithDescriptionAndDetail(_, desc, ref detail) => {
                desc.fmt(f)?;
                f.write_str(": ")?;
                detail.fmt(f)
            }
            ErrorRepr::ExtensionError(ref code, ref detail) => {
                code.fmt(f)?;
                f.write_str(": ")?;
                detail.fmt(f)
            }
            ErrorRepr::IoError(ref err) => err.fmt(f),
        }
    }
}

impl fmt::Debug for RpError {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        fmt::Display::fmt(self, f)
    }
}

/// Indicates a general failure in the library.
impl RpError {
    /// Returns the kind of the error.
    pub fn kind(&self) -> ErrorKind {
        match self.repr {
            ErrorRepr::WithDescription(kind, _) => kind,
            ErrorRepr::WithDescriptionAndDetail(kind, _, _) => kind,
            ErrorRepr::ExtensionError(_, _) => ErrorKind::ExtensionError,
            ErrorRepr::IoError(_) => ErrorKind::IoError,
        }
    }

    /// Returns the name of the error category for display purposes.
    pub fn category(&self) -> &str {
        match self.kind() {
            ErrorKind::NoLeftSpaceError => "no left space error",
            ErrorKind::BufferOverMaxError => "buffer over max error",
            ErrorKind::TypeNotMatchError => "type not match error",
            ErrorKind::ParseError => "parse error",
            ErrorKind::MissingError => "missing error",
            ErrorKind::StringFormatError => "string format error",
            ErrorKind::IoError => "I/O error",
            ErrorKind::ExtensionError => "extension error",
        }
    }

    /// Indicates that this failure is an IO failure.
    pub fn is_io_error(&self) -> bool {
        match self.kind() {
            ErrorKind::IoError => true,
            _ => false,
        }
    }

    /// Returns the extension error code
    pub fn extension_error_code(&self) -> Option<&str> {
        match self.repr {
            ErrorRepr::ExtensionError(ref code, _) => Some(&code),
            _ => None,
        }
    }

    /// Returns the extension error detail
    pub fn extension_error_detail(&self) -> Option<&str> {
        match self.repr {
            ErrorRepr::ExtensionError(_, ref detail) => Some(&detail),
            ErrorRepr::WithDescriptionAndDetail(_, _, ref detail) => Some(&detail),
            _ => None,
        }
    }
}

pub fn make_extension_error(code: &str, detail: Option<&str>) -> RpError {
    RpError {
        repr: ErrorRepr::ExtensionError(code.to_string(),
                                        match detail {
                                            Some(x) => x.to_string(),
                                            None => {
                                                "Unknown extension error encountered".to_string()
                                            }
                                        }),
    }
}

File: dotnet/src/Values.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Collections;

namespace proto.tunm
{

    public class TunmValues
    {
        public const byte TYPE_NIL = 0;
        public const byte TYPE_BOOL = 1;
        public const byte TYPE_U8 = 2;
        public const byte TYPE_I8 = 3;
        public const byte TYPE_U16 = 4;
        public const byte TYPE_I16 = 5;
        public const byte TYPE_U32 = 6;
        public const byte TYPE_I32 = 7;
        public const byte TYPE_U64 = 8;
        public const byte TYPE_I64 = 9;
        public const byte TYPE_VARINT = 10;
        public const byte TYPE_FLOAT = 11;
        public const byte TYPE_DOUBLE = 12;
        public const byte TYPE_STR = 13;
        public const byte TYPE_STR_IDX = 14;
        public const byte TYPE_RAW = 15;
        public const byte TYPE_ARR = 16;
        public const byte TYPE_MAP = 17;

        public const Object NilObject = null;

        public const string STR_TYPE_NIL = "Null";
        public const string STR_TYPE_BOOL = "Boolean";
        public const string STR_TYPE_U8 = "Byte";
        public const string STR_TYPE_I8 = "IByte";
        public const string STR_TYPE_U16 = "UInt16";
        public const string STR_TYPE_I16 = "Int16";
        public const string STR_TYPE_U32 = "UInt32";
        public const string STR_TYPE_I32 = "Int32";
        public const string STR_TYPE_U64 = "UInt64";
        public const string STR_TYPE_I64 = "Int64";
        public const string STR_TYPE_VARINT = "VARINT";
        public const string STR_TYPE_FLOAT = "Single";
        public const string STR_TYPE_DOUBLE = "Double";
        public const string STR_TYPE_STR = "String";
        public const string STR_TYPE_STR_IDX = "StringIdx";
        public const string STR_TYPE_RAW = "Byte[]";
        public const string STR_TYPE_ARR = "ArrayList";
        public const string STR_TYPE_MAP = "Dictionary`2";

        static public byte get_type_by_value(ref Object obj)
        {
            if (obj == null)
            {
                return TYPE_NIL;
            }
            return get_type_by_name(obj.GetType().Name);
        }

        static public byte get_type_by_name(string name)
        {
            switch (name)
            {
                case STR_TYPE_NIL: return TunmValues.TYPE_NIL;
                case STR_TYPE_BOOL: return TunmValues.TYPE_BOOL;
                case STR_TYPE_U8: return TunmValues.TYPE_U8;
                case STR_TYPE_I8: return TunmValues.TYPE_I8;
                case STR_TYPE_U16: return TunmValues.TYPE_U16;
                case STR_TYPE_I16: return TunmValues.TYPE_I16;
                case STR_TYPE_U32: return TunmValues.TYPE_U32;
                case STR_TYPE_I32: return TunmValues.TYPE_I32;
                case STR_TYPE_U64: return TunmValues.TYPE_U64;
                case STR_TYPE_I64: return TunmValues.TYPE_I64;
                case STR_TYPE_VARINT: return TunmValues.TYPE_VARINT;
                case STR_TYPE_FLOAT: return TunmValues.TYPE_FLOAT;
                case STR_TYPE_DOUBLE: return TunmValues.TYPE_DOUBLE;
                case STR_TYPE_STR: return TunmValues.TYPE_STR;
                case STR_TYPE_STR_IDX: return TunmValues.TYPE_STR_IDX;
                case STR_TYPE_RAW: return TunmValues.TYPE_RAW;
                case STR_TYPE_ARR: return TunmValues.TYPE_ARR;
                case STR_TYPE_MAP: return TunmValues.TYPE_MAP;
                default: return TunmValues.TYPE_NIL;
            }
        }

        static public string get_name_by_type(byte index)
        {
            switch (index)
            {
                case TunmValues.TYPE_NIL: return TunmValues.STR_TYPE_NIL;
                case TunmValues.TYPE_BOOL: return TunmValues.STR_TYPE_BOOL;
                case TunmValues.TYPE_U8: return TunmValues.STR_TYPE_U8;
                case TunmValues.TYPE_I8: return TunmValues.STR_TYPE_I8;
                case TunmValues.TYPE_U16: return TunmValues.STR_TYPE_U16;
                case TunmValues.TYPE_I16: return TunmValues.STR_TYPE_I16;
                case TunmValues.TYPE_U32: return TunmValues.STR_TYPE_U32;
                case TunmValues.TYPE_I32: return TunmValues.STR_TYPE_I32;
                case TunmValues.TYPE_U64: return TunmValues.STR_TYPE_U64;
                case TunmValues.TYPE_I64: return TunmValues.STR_TYPE_I64;
                case TunmValues.TYPE_VARINT: return TunmValues.STR_TYPE_VARINT;
                case TunmValues.TYPE_FLOAT: return TunmValues.STR_TYPE_FLOAT;
                case TunmValues.TYPE_DOUBLE: return TunmValues.STR_TYPE_DOUBLE;
                case TunmValues.TYPE_STR: return TunmValues.STR_TYPE_STR;
                case TunmValues.TYPE_STR_IDX: return TunmValues.STR_TYPE_STR_IDX;
                case TunmValues.TYPE_RAW: return TunmValues.STR_TYPE_RAW;
                case TunmValues.TYPE_ARR: return TunmValues.STR_TYPE_ARR;
                case TunmValues.TYPE_MAP: return TunmValues.STR_TYPE_MAP;
                default: return TunmValues.STR_TYPE_NIL;
            }
        }


        public static String to_string(object obj)
        {
            var builder = new StringBuilder();
            to_string(ref builder, obj);
            return builder.ToString();
        }

        public static void to_string(ref StringBuilder builder, object obj)
        {
            if (obj == null)
            {
                builder.Append("null");
                return;
            }
            var name = obj.GetType().Name;
            if (name == TunmValues.STR_TYPE_STR)
            {
                builder.AppendFormat("\"{0}\"", (String)obj);
            }
            else if (name == TunmValues.STR_TYPE_RAW)
            {
                builder.AppendFormat("\"byte[]({0})\"", ((byte[])obj).Length);
            }
            else if (name == TunmValues.STR_TYPE_ARR)
            {
                var list = (ArrayList)(obj);
                var index = 0;
                builder.Append("[");
                foreach (var sub_val in list)
                {
                    to_string(ref builder, sub_val);
                    if (++index < list.Count)
                    {
                        builder.Append(",");
                    }
                }
                builder.Append("]");
            }
            else if (name == TunmValues.STR_TYPE_MAP)
            {
                var index = 0;
                var map = (Dictionary<object, object>)(obj);
                builder.Append("{");
                foreach (var key in map)
                {
                    to_string(ref builder, key.Key);
                    builder.Append(":");
                    to_string(ref builder, key.Value);
                    if (++index < map.Count)
                    {
                        builder.Append(",");
                    }
                }
                builder.Append("}");
            }
            else if (name == TunmValues.STR_TYPE_MAP)
            {
                var map = (Dictionary<object, object>)(obj);
                builder.AppendFormat("Dictionary{}{", map.Count);
                foreach (var key in map)
                {
                    to_string(ref builder, key);
                    builder.Append(":");
                    to_string(ref builder, map[key]);
                    builder.Append(",");
                }
                builder.Append("}");
            }
            else
            {
                switch (name)
                {
                    case TunmValues.STR_TYPE_I8:
                    case TunmValues.STR_TYPE_U8:
                        builder.Append((byte)obj);
                        break;
                    case TunmValues.STR_TYPE_I16:
                        builder.Append((short)obj);
                        break;
                    case TunmValues.STR_TYPE_U16:
                        builder.Append((ushort)obj);
                        break;
                    case TunmValues.STR_TYPE_I32:
                        builder.Append((int)obj);
                        break;
                    case TunmValues.STR_TYPE_U32:
                        builder.Append((uint)obj);
                        break;
                    case TunmValues.STR_TYPE_I64:
                        builder.Append((long)obj);
                        break;
                    case TunmValues.STR_TYPE_U64:
                        builder.Append((ulong)obj);
                        break;
                    case TunmValues.STR_TYPE_FLOAT:
                        {
                            builder.Append((float)obj);
                            break;
                        }
                    case TunmValues.STR_TYPE_DOUBLE:
                        {
                            builder.Append((double)obj);
                            break;
                        }
                }
            }
        }

        public static bool is_string(Object obj)
        {
            if (obj == null)
            {
                return false;
            }
            if (obj.GetType().Name == STR_TYPE_STR)
            {
                return true;
            }
            return false;
        }

        public static bool is_arr(Object obj)
        {
            if (obj == null)
            {
                return false;
            }
            if (obj.GetType().Name == STR_TYPE_ARR)
            {
                return true;
            }
            return false;
        }

        public static bool is_dictionary(Object obj)
        {
            if (obj == null)
            {
                return false;
            }
            if (obj.GetType().Name == STR_TYPE_MAP)
            {
                return true;
            }
            return false;
        }

        public static bool is_long(Object obj)
        {
            if (obj == null)
            {
                return false;
            }

            switch (obj.GetType().Name)
            {
                case TunmValues.STR_TYPE_I8:
                case TunmValues.STR_TYPE_U8:
                    return true;
                case TunmValues.STR_TYPE_I16:
                    return true;
                case TunmValues.STR_TYPE_U16:
                    return true;
                case TunmValues.STR_TYPE_I32:
                    return true;
                case TunmValues.STR_TYPE_U32:
                    return true;
                case TunmValues.STR_TYPE_I64:
                    return true;
                case TunmValues.STR_TYPE_U64:
                    return true;
            }
            return false;
        }

        public static bool is_double(Object obj)
        {
            if (obj == null)
            {
                return false;
            }

            switch (obj.GetType().Name)
            {
                case TunmValues.STR_TYPE_FLOAT:
                    return true;
                case TunmValues.STR_TYPE_DOUBLE:
                    return true;
            }
            return false;
        }

        public static String get_string(Object obj)
        {
            if (is_string(obj))
            {
                return (String)(obj);
            }
            return "";
        }

        public static Dictionary<Object, Object> get_dictionary(Object obj)
        {
            if (is_dictionary(obj))
            {
                return (Dictionary<Object, Object>)(obj);
            }
            return new Dictionary<Object, Object>();
        }

        public static ArrayList get_arr(Object obj)
        {
            if (is_arr(obj))
            {
                return (ArrayList)(obj);
            }
            return new ArrayList();
        }

        public static long get_long(Object obj)
        {
            long val = 0;
            switch (obj.GetType().Name)
            {
                case TunmValues.STR_TYPE_I8:
                case TunmValues.STR_TYPE_U8:
                    val = (byte)obj;
                    break;
                case TunmValues.STR_TYPE_I16:
                    val = (short)obj;
                    break;
                case TunmValues.STR_TYPE_U16:
                    val = (ushort)obj;
                    break;
                case TunmValues.STR_TYPE_I32:
                    val = (int)obj;
                    break;
                case TunmValues.STR_TYPE_U32:
                    val = (uint)obj;
                    break;
                case TunmValues.STR_TYPE_I64:
                    val = (long)obj;
                    break;
                case TunmValues.STR_TYPE_U64:
                    val = (long)(ulong)obj;
                    break;
            }
            return val;
        }

        public static int get_int(Object obj)
        {
            return Convert.ToInt32(get_long(obj));
        }

        public static double get_double(Object obj)
        {
            double val = 0;
            switch (obj.GetType().Name)
            {
                case TunmValues.STR_TYPE_FLOAT:
                    val = (float)obj;
                    break;
                case TunmValues.STR_TYPE_DOUBLE:
                    val = (double)obj;
                    break;
            }
            return val;
        }


        public static String get_string(Object obj, String key)
        {
            var map = get_dictionary(obj);
            if (map == null)
            {
                return "";
            }
            if (map.ContainsKey(key))
            {
                return get_string(map[key]);
            }
            return "";
        }

        public static String get_string(Object obj, int index)
        {
            var arr = get_arr(obj);
            
            if (arr == null)
            {
                return "";
            }

            if (arr.Count > index)
            {
                return get_string(arr[index]);
            }
            return "";
        }
        
        public static byte[] get_bytes(Object obj,string key){
            var map = get_dictionary(obj);
            if (map == null)
            {
                return null;
            }
            if (map.ContainsKey(key))
            {
                return (byte[])(map[key]);
            }
            return new byte[0];
        }


        public static long get_long(Object obj, String key)
        {
            var map = get_dictionary(obj);
            if (map == null)
            {
                return 0;
            }
            if (map.ContainsKey(key))
            {
                return get_long(map[key]);
            }
            return 0;
        }

        public static long get_long(Object obj, int index)
        {
            var arr = get_arr(obj);
            if (arr == null)
            {
                return 0;
            }

            if (arr.Count > index)
            {
                return get_long(arr[index]);
            }
            return 0;
        }


        public static int get_int(Object obj, String key)
        {
            return Convert.ToInt32(get_long(obj, key));
        }

        public static int get_int(Object obj, int index)
        {
            return Convert.ToInt32(get_long(obj, index));
        }


        public static double get_double(Object obj, String key)
        {
            var map = get_dictionary(obj);
            if (map == null)
            {
                return 0;
            }
            if (map.ContainsKey(key))
            {
                return get_double(map[key]);
            }
            return 0;
        }

        public static double get_double(Object obj, int index)
        {
            var arr = get_arr(obj);
            if (arr == null)
            {
                return 0;
            }

            if (arr.Count > index)
            {
                return get_double(arr[index]);
            }
            return 0;
        }


        public static ArrayList get_arr(Object obj, String key)
        {
            var map = get_dictionary(obj);
            if (map == null)
            {
                return null;
            }
            if (map.ContainsKey(key))
            {
                return get_arr(map[key]);
            }
            return new ArrayList();
        }

        public static ArrayList get_arr(Object obj, int index)
        {
            var arr = get_arr(obj);
            if (arr == null)
            {
                return new ArrayList();
            }

            if (arr.Count > index)
            {
                return get_arr(arr[index]);
            }
            return new ArrayList();
        }


        public static Dictionary<Object, Object> get_dictionary(Object obj, String key)
        {
            var map = get_dictionary(obj);
            if (map == null)
            {
                return new Dictionary<Object, Object>();
            }
            if (map.ContainsKey(key))
            {
                return get_dictionary(map[key]);
            }
            return new Dictionary<Object, Object>();
        }

        public static Dictionary<Object, Object> get_dictionary(Object obj, int index)
        {
            var arr = get_arr(obj);
            if (arr == null)
            {
                return new Dictionary<Object, Object>();
            }

            if (arr.Count > index)
            {
                return get_dictionary(arr[index]);
            }
            return new Dictionary<Object, Object>();
        }
    }
}

